ENCD-5100-index-reads-grouping

forresttanaka opened the pull request:
Nearly all the relevant change is in filegallery.js. I had to add a lot of local test data to be able to develop this. I have a demo with this current build, with a search that shows the paired-ended case. We have no single-ended cases in production data, but I have one in local test data.

forresttanaka committed:
ENCD-5100-index-reads-grouping

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -151,7 +151,13 @@ export class FileTable extends React.Component {
             const files = _(datasetFiles).groupBy((file) => {
                 if (file.output_category === 'raw data') {
-                    return file.output_type === 'reads' ? 'raw' : 'rawArray';
+                    if (file.output_type === 'reads') {
```
Would an if- else if be a tad better? Isolated if-statements imply the if-statements are not related nor operating on the same variable.

forresttanaka replied:
With embedded returns, that wouldn’t be allowed under [AirBnB](https://github.com/airbnb/javascript#arrays--callback-return).

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -482,16 +489,30 @@ class RawSequencingTable extends React.Component {
         };
 
         // Bind `this` to non-React methods.
+        this.findIndexFile = this.findIndexFile.bind(this);
```
findIndexFile or GetIndexFile/FetchIndexFile?

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -482,16 +489,30 @@ class RawSequencingTable extends React.Component {
+     *
+     * @return {object} Index reads file with index_of matching given files; undefined if none
+     */
+    findIndexFile(file0, file1) {
```
This is a find. However, from the POV of the consumer of the method, it is a GET

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -528,24 +549,61 @@ class RawSequencingTable extends React.Component {
-                            file.pairSortKey += file.paired_end;
-                            partner.pairSortKey += partner.paired_end;
+                            // Both the file and its partner qualify as good pairs of each other.
+                            // Let hem pass the filter, and set their sort keys to the lower of the
```
"them" and not "hem"

forresttanaka replied:
I’m going to change that..

zoldello replied:
did you change this?

forresttanaka replied:
Yes. You’re looking at the outdated version.

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -505,8 +526,8 @@ class RawSequencingTable extends React.Component {
             // paired.
-            const nonpairedFiles = [];
-            const pairedFiles = _(files).filter((file) => {
+            let nonpairedFiles = [];
```
Why did you change this from "const" to "let"? ESLINT did not bark at me when pulled down this code and changed these to const locally.

forresttanaka replied:
nonpairedFiles gets reassigned on line 573. ESLint gives me an error.

zoldello replied:
Thanks. I missed that.

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -528,24 +549,61 @@ class RawSequencingTable extends React.Component {
-                nonpairedFiles.push(file);
+                // File not part of a pair; add it to the non-paired list as a copy so we can
+                // mutate them later.
+                nonpairedFiles.push(Object.assign({}, file));
```
Is there a reason why you created a new object via Object.assign. rather than use file-object itself?

It would be cool if we upgrade babel to support object spread (https://stackoverflow.com/q/32925460/178550)

forresttanaka replied:
I’m mutating the file objects, and the file objects come in as a prop.

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -528,24 +549,61 @@ class RawSequencingTable extends React.Component {
+                    const matchingPairFile = pairedFiles.find(pairedFile => indexFile.index_of.includes(pairedFile['@id']));
+                    if (matchingPairFile) {
+                        indexFileCopy.pairSortId = matchingPairFile.pairSortId;
+                        indexFileCopy.pairSortKey = `${matchingPairFile.pairSortId}-I`;
```
Is the 'I' at the end intentional?

forresttanaka replied:
Very much, so that it always sorts behind {accession}-1 and {accession}-2. I never explain that though, and I should.

zoldello commented on src/encoded/tests/test_batch_download.py:
```diff
@@ -328,7 +328,7 @@ def test_batch_download_report_download(testapp, workbook):
         b'Replicates',
     ]
-    assert len(lines) == 64
+    assert len(lines) == 65
```
This increased by 1 because you added a new experiments. Now, what is special about the one you added? The data looks like the others. Was it "alias"?

forresttanaka replied:
Because I didn’t want to use another experiment to test specifically one with index reads files, in case that removed a test of something else.

zoldello commented on src/encoded/static/components/filegallery.js:
```diff
@@ -554,12 +612,12 @@ class RawSequencingTable extends React.Component {
-                        'Z'
-                ));
+                pairedRepGroups = _(pairedFiles).groupBy((file) => {
+                    if (file.biological_replicates && file.biological_replicates.length === 1) {
```
How is this an improvement from the previous code?

forresttanaka replied:
I had initially modified this to do additional things and had to remove the ternary expression. I removed that change, but felt the ternary condition was so long and unwieldy, I preferred using the if.

zoldello reviewed (dismissed):
The code looks fine. My comments are nothing that must be changed. I am not sure if it is worth even changing things for them small typo

forresttanaka committed:
Update comments

zoldello reviewed (dismissed):
Looks good.

forresttanaka committed:
Change index reads display

Remove “SE” from run type for index-reads files associated with single-ended files.

zoldello reviewed (approved):
Updated (rawIsIndex) looks fine.
